src/OpenSSL/crypto.py

```diff
@@ -2583,9 +2583,9 @@ def sign(pkey, data, digest):
     _lib.EVP_SignInit(md_ctx, digest_obj)
     _lib.EVP_SignUpdate(md_ctx, data, len(data))
 
-    signature_buffer = _ffi.new("unsigned char[]", 512)
+    pkey_length = PKey.bits(pkey) // 8
```

Comment:
I think probably you want `pkey_length = (pkey.bits() + 7) // 8`, otherwise rounding down will get you for a key that's not a multiple of 8 bits.